### PR TITLE
[FOLLOWUP] Enforcement Rule: resolve review comments, refactor adjust_input_keys_ordering()

### PR DESCRIPTION
**Pull request by mingmwang**
# Which issue does this PR close?

Follow up to https://github.com/apache/arrow-datafusion/pull/4122

# Rationale for this change

# What changes are included in this PR?

1)  Add more comments and examples to explain the rationale and process of join key reordering
2)  Refactor `adjust_input_keys_ordering()`, leverage the `transform_down() `method to drive the Top-Down process
3) UT for `Partition` satisfy `Distribution`

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by mingmwang** [comment]
@alamb @yahoNanJing @jackwener 
Please help to take a look.

File: datafusion/core/src/physical_optimizer/merge_exec.rs [hunk 1]
```diff
@@ -1,75 +0,0 @@
-// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/merge_exec.rs:1]: `merge_exec` is a strange name for this module -- I am glad we got rid of it!

File: datafusion/core/src/execution/context.rs [hunk 2]
```diff
@@ -1589,8 +1589,9 @@ impl SessionState {
             )));
         }
         physical_optimizers.push(Arc::new(Repartition::new()));
+        // Repartition rule could introduce additional RepartitionExec with RoundRobin partitioning.
+        // To make sure the SinglePartition is satisfied, run the BasicEnforcement again, originally it was the AddCoalescePartitionsExec here.
```
Comment (alamb) [inline datafusion/core/src/execution/context.rs:1593]: 👍

File: datafusion/core/src/physical_plan/mod.rs [hunk 3]
```diff
@@ -656,3 +657,81 @@ use crate::execution::context::TaskContext;
...
+            Distribution::HashPartitioned(partition_exprs1.clone()),
+        ];
+
+        let single_partition = Partitioning::UnknownPartitioning(1);
+        let unspecified_partition = Partitioning::UnknownPartitioning(10);
+        let round_robin_partition = Partitioning::RoundRobinBatch(10);
+        let hash_partition1 = Partitioning::Hash(partition_exprs1, 10);
+        let hash_partition2 = Partitioning::Hash(partition_exprs2, 10);
+
+        for distribution in distribution_types {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/mod.rs:703]: thank you

File: datafusion/core/src/physical_plan/rewrite.rs [hunk 4]
```diff
@@ -149,14 +149,16 @@ pub enum RewriteRecursion {
     Skip,
 }
 
+#[allow(clippy::vtable_address_comparisons)]
```
Comment (alamb) [inline datafusion/core/src/physical_plan/rewrite.rs:152]: Do you know why  `[vtable](clippy::vtable_address_comparisons)` is needed here but not in the seemingly very similar other changes in this PR?

File: datafusion/physical-expr/src/equivalence.rs [hunk 5]
```diff
@@ -247,11 +247,16 @@ mod tests {
...
         assert_eq!(eq_properties.classes().len(), 1);
         assert_eq!(eq_properties.classes()[0].len(), 2);
+        assert!(eq_properties.classes()[0].contains(&Column::new("a", 0)));
+        assert!(eq_properties.classes()[0].contains(&Column::new("b", 1)));
 
         let new_condition = (&Column::new("b", 1), &Column::new("c", 2));
         eq_properties.add_equal_conditions(new_condition);
         assert_eq!(eq_properties.classes().len(), 1);
         assert_eq!(eq_properties.classes()[0].len(), 3);
+        assert!(eq_properties.classes()[0].contains(&Column::new("a", 0)));
```
Comment (alamb) [inline datafusion/physical-expr/src/equivalence.rs:257]: Thank you

File: datafusion/core/src/physical_optimizer/enforcement.rs [hunk 6]
```diff
@@ -99,13 +105,51 @@ impl PhysicalOptimizerRule for BasicEnforcement {
...
 /// When the physical planner creates the Joins, the ordering of join keys is from the original query.
 /// That might not match with the output partitioning of the join node's children
-/// This method runs a top-down process and try to adjust the output partitioning of the children
-/// if the children themselves are Joins or Aggregations.
-fn adjust_input_keys_down_recursively(
-    plan: Arc<dyn crate::physical_plan::ExecutionPlan>,
-    parent_required: Vec<Arc<dyn PhysicalExpr>>,
-) -> Result<Arc<dyn crate::physical_plan::ExecutionPlan>> {
-    let plan_any = plan.as_any();
+/// A Top-Down process will use this method to adjust children's output partitioning based on the parent key reordering requirements:
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/enforcement.rs:108]: ❤️

File: datafusion/core/src/physical_optimizer/enforcement.rs [hunk 7]
```diff
@@ -885,6 +956,70 @@ struct JoinKeyPairs {
     right_keys: Vec<Arc<dyn PhysicalExpr>>,
 }
 
+#[derive(Debug, Clone)]
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/enforcement.rs:959]: Nice

**Review by alamb — APPROVED** [review]
Thank you very much for the follow up @mingmwang  🙏  This looks great

(for other reviewers, I found whitespace blind diff https://github.com/apache/arrow-datafusion/pull/4184/files?w=1 easier to read and understand)

**Review by jackwener — APPROVED** [review]
LGTM

File: datafusion/core/src/physical_plan/rewrite.rs (hunk 4 above)
Comment (mingmwang) [reply datafusion/core/src/physical_plan/rewrite.rs:152]: Sorry, this is useless. I will remove it from the PR.

**Pull request merged**